File: app/src/main/java/com/pdfimport/webview/MainActivity.java
package com.pdfimport.webview;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.webkit.JavascriptInterface;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.annotation.SuppressLint;
import android.util.Base64;
import android.util.Log;
import android.widget.Toast;

import java.io.ByteArrayOutputStream;
import java.io.InputStream;

public class MainActivity extends Activity {
    private WebView webView;
    private static final int PICK_PDF_REQUEST = 1;
    private static final String TAG = "PDFImport";
    private String pendingFileData = null;
    
    @SuppressLint({"SetJavaScriptEnabled", "JavascriptInterface"})
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        
        webView = new WebView(this);
        webView.setWebViewClient(new WebViewClient());
        
        WebSettings settings = webView.getSettings();
        settings.setJavaScriptEnabled(true);
        settings.setAllowFileAccess(true);
        settings.setAllowContentAccess(true);
        settings.setDomStorageEnabled(true);
        
        // إضافة JavaScript Interface للتواصل مع HTML
        webView.addJavascriptInterface(new JavaScriptInterface(), "AndroidBridge");
        
        webView.loadUrl("file:///android_asset/www/index.html");
        
        setContentView(webView);
    }
    
    // JavaScript Interface للتواصل مع HTML
    private class JavaScriptInterface {
        @JavascriptInterface
        public void openFilePicker() {
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    Intent intent = new Intent(Intent.ACTION_GET_CONTENT);
                    intent.addCategory(Intent.CATEGORY_OPENABLE);
                    intent.setType("application/pdf");
                    startActivityForResult(Intent.createChooser(intent, "Select PDF"), PICK_PDF_REQUEST);
                }
            });
        }
        
        @JavascriptInterface
        public String getSelectedFile() {
            return pendingFileData;
        }
        
        @JavascriptInterface
        public void clearSelectedFile() {
            pendingFileData = null;
        }
    }
    
    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        
        if (requestCode == PICK_PDF_REQUEST && resultCode == Activity.RESULT_OK) {
            if (data != null && data.getData() != null) {
                try {
                    Uri uri = data.getData();
                    byte[] fileBytes = readBytes(uri);
                    String base64 = Base64.encodeToString(fileBytes, Base64.NO_WRAP);
                    pendingFileData = base64;
                    
                    // إرسال الملف إلى JavaScript
                    String js = "javascript:receiveFile('" + base64 + "')";
                    webView.post(new Runnable() {
                        @Override
                        public void run() {
                            webView.evaluateJavascript(js, null);
                        }
                    });
                    
                } catch (Exception e) {
                    Log.e(TAG, "Error reading file", e);
                    Toast.makeText(this, "Error: " + e.getMessage(), Toast.LENGTH_LONG).show();
                }
            }
        }
    }
    
    private byte[] readBytes(Uri uri) throws Exception {
        InputStream inputStream = getContentResolver().openInputStream(uri);
        ByteArrayOutputStream byteBuffer = new ByteArrayOutputStream();
        
        byte[] buffer = new byte[1024 * 1024]; // 1MB buffer
        int len;
        while ((len = inputStream.read(buffer)) != -1) {
            byteBuffer.write(buffer, 0, len);
        }
        
        return byteBuffer.toByteArray();
    }
    
    @Override
    public void onBackPressed() {
        if (webView.canGoBack()) {
            webView.goBack();
        } else {
            super.onBackPressed();
        }
    }
}

File: app/src/main/java/com/pdfimport/app/MainActivity.kt
package com.pdfimport.app

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.provider.OpenableColumns
import android.widget.Button
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import com.chaquo.python.Python
import com.chaquo.python.android.AndroidPlatform
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileOutputStream

class MainActivity : AppCompatActivity() {
    
    private lateinit var progressBar: ProgressBar
    private lateinit var statusText: TextView
    private lateinit var selectButton: Button
    private lateinit var processButton: Button
    
    private var selectedPdfUri: Uri? = null
    private var selectedPdfPath: String? = null
    private var selectedFileName: String = "document.pdf"
    
    private val PICK_PDF_REQUEST = 1
    
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        
        initializeViews()
        initializePython()
        setupListeners()
    }
    
    private fun initializeViews() {
        progressBar = findViewById(R.id.progressBar)
        statusText = findViewById(R.id.statusText)
        selectButton = findViewById(R.id.selectButton)
        processButton = findViewById(R.id.processButton)
        
        progressBar.visibility = ProgressBar.INVISIBLE
        processButton.isEnabled = false
    }
    
    private fun initializePython() {
        if (!Python.isStarted()) {
            Python.start(AndroidPlatform(this))
        }
    }
    
    private fun setupListeners() {
        selectButton.setOnClickListener {
            openFilePicker()
        }
        
        processButton.setOnClickListener {
            selectedPdfPath?.let { path ->
                processPDF(path)
            }
        }
    }
    
    private fun openFilePicker() {
        val intent = Intent(Intent.ACTION_OPEN_DOCUMENT).apply {
            addCategory(Intent.CATEGORY_OPENABLE)
            type = "application/pdf"
        }
        startActivityForResult(intent, PICK_PDF_REQUEST)
    }
    
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        
        if (requestCode == PICK_PDF_REQUEST && resultCode == RESULT_OK) {
            data?.data?.let { uri ->
                selectedPdfUri = uri
                selectedFileName = getFileName(uri)
                copyPdfToInternalStorage(uri) { path ->
                    selectedPdfPath = path
                    runOnUiThread {
                        processButton.isEnabled = true
                        statusText.text = "PDF selected: $selectedFileName"
                        statusText.setTextColor(getColor(android.R.color.holo_green_dark))
                    }
                }
            }
        }
    }
    
    private fun copyPdfToInternalStorage(uri: Uri, callback: (String) -> Unit) {
        lifecycleScope.launch(Dispatchers.IO) {
            try {
                val inputStream = contentResolver.openInputStream(uri)
                val outputFile = File(cacheDir, selectedFileName)
                
                inputStream?.use { input ->
                    FileOutputStream(outputFile).use { output ->
                        input.copyTo(output)
                    }
                }
                
                withContext(Dispatchers.Main) {
                    callback(outputFile.absolutePath)
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    Toast.makeText(this@MainActivity, "Error copying file: ${e.message}", Toast.LENGTH_LONG).show()
                    statusText.text = "Error: ${e.message}"
                    statusText.setTextColor(getColor(android.R.color.holo_red_dark))
                }
            }
        }
    }
    
    private fun getFileName(uri: Uri): String {
        var fileName = "document.pdf"
        contentResolver.query(uri, null, null, null, null)?.use { cursor ->
            if (cursor.moveToFirst()) {
                val nameIndex = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME)
                if (nameIndex >= 0) {
                    fileName = cursor.getString(nameIndex)
                }
            }
        }
        return fileName
    }
    
    private fun processPDF(pdfPath: String) {
        lifecycleScope.launch {
            try {
                progressBar.visibility = ProgressBar.VISIBLE
                progressBar.isIndeterminate = true
                statusText.text = "Processing PDF..."
                statusText.setTextColor(getColor(android.R.color.holo_blue_dark))
                processButton.isEnabled = false
                selectButton.isEnabled = false
                
                val result = withContext(Dispatchers.IO) {
                    PythonPDFProcessor.processPDF(pdfPath)
                }
                
                progressBar.visibility = ProgressBar.INVISIBLE
                statusText.text = "Processing complete!"
                statusText.setTextColor(getColor(android.R.color.holo_green_dark))
                processButton.isEnabled = true
                selectButton.isEnabled = true
                
                showResults(result)
            } catch (e: Exception) {
                progressBar.visibility = ProgressBar.INVISIBLE
                statusText.text = "Error: ${e.message}"
                statusText.setTextColor(getColor(android.R.color.holo_red_dark))
                processButton.isEnabled = true
                selectButton.isEnabled = true
                Toast.makeText(this@MainActivity, "Processing error: ${e.message}", Toast.LENGTH_LONG).show()
            }
        }
    }
    
    private fun showResults(result: PDFProcessingResult) {
        val intent = Intent(this, ResultsActivity::class.java)
        intent.putExtra("result_json", result.toJson())
        startActivity(intent)
    }
}
